<h3>Exercise 1: Loading Data with Pandas</h3>


In [1]:
import pandas as pd

In [4]:
df = pd.read_csv("framingham.csv")

In [5]:
df

male  age  education  currentSmoker  cigsPerDay  BPMeds  \
0        1   39        4.0              0         0.0     0.0   
1        0   46        2.0              0         0.0     0.0   
2        1   48        1.0              1        20.0     0.0   
3        0   61        3.0              1        30.0     0.0   
4        0   46        3.0              1        23.0     0.0   
...    ...  ...        ...            ...         ...     ...   
4233     1   50        1.0              1         1.0     0.0   
4234     1   51        3.0              1        43.0     0.0   
4235     0   48        2.0              1        20.0     NaN   
4236     0   44        1.0              1        15.0     0.0   
4237     0   52        2.0              0         0.0     0.0   

      prevalentStroke  prevalentHyp  diabetes  totChol  sysBP  diaBP    BMI  \
0                   0             0         0    195.0  106.0   70.0  26.97   
1                   0             0         0    250.0  121.0   81.0  28.73   
2                   0             0         0    245.0  127.5   80.0  25.34   
3                   0             1         0    225.0  150.0   95.0  28.58   
4                   0             0         0    285.0  130.0   84.0  23.10   
...               ...           ...       ...      ...    ...    ...    ...   
4233                0             1         0    313.0  179.0   92.0  25.97   
4234                0             0         0    207.0  126.5   80.0  19.71   
4235                0             0         0    248.0  131.0   72.0  22.00   
4236                0             0         0    210.0  126.5   87.0  19.16   
4237                0             0         0    269.0  133.5   83.0  21.47   

      heartRate  glucose  TenYearCHD  
0          80.0     77.0           0  
1          95.0     76.0           0  
2          75.0     70.0           0  
3          65.0    103.0           1  
4          85.0     85.0           0  
...         ...      ...         ...  
4233       66.0     86.0           1  
4234       65.0     68.0           0  
4235       84.0     86.0           0  
4236       86.0      NaN           0  
4237       80.0    107.0           0  

[4238 rows x 16 columns]

In [6]:
df.head()

male  age  education  currentSmoker  cigsPerDay  BPMeds  prevalentStroke  \
0     1   39        4.0              0         0.0     0.0                0   
1     0   46        2.0              0         0.0     0.0                0   
2     1   48        1.0              1        20.0     0.0                0   
3     0   61        3.0              1        30.0     0.0                0   
4     0   46        3.0              1        23.0     0.0                0   

   prevalentHyp  diabetes  totChol  sysBP  diaBP    BMI  heartRate  glucose  \
0             0         0    195.0  106.0   70.0  26.97       80.0     77.0   
1             0         0    250.0  121.0   81.0  28.73       95.0     76.0   
2             0         0    245.0  127.5   80.0  25.34       75.0     70.0   
3             1         0    225.0  150.0   95.0  28.58       65.0    103.0   
4             0         0    285.0  130.0   84.0  23.10       85.0     85.0   

   TenYearCHD  
0           0  
1           0  
2           0  
3           1  
4           0

In [7]:
df.tail()

male  age  education  currentSmoker  cigsPerDay  BPMeds  \
4233     1   50        1.0              1         1.0     0.0   
4234     1   51        3.0              1        43.0     0.0   
4235     0   48        2.0              1        20.0     NaN   
4236     0   44        1.0              1        15.0     0.0   
4237     0   52        2.0              0         0.0     0.0   

      prevalentStroke  prevalentHyp  diabetes  totChol  sysBP  diaBP    BMI  \
4233                0             1         0    313.0  179.0   92.0  25.97   
4234                0             0         0    207.0  126.5   80.0  19.71   
4235                0             0         0    248.0  131.0   72.0  22.00   
4236                0             0         0    210.0  126.5   87.0  19.16   
4237                0             0         0    269.0  133.5   83.0  21.47   

      heartRate  glucose  TenYearCHD  
4233       66.0     86.0           1  
4234       65.0     68.0           0  
4235       84.0     86.0           0  
4236       86.0      NaN           0  
4237       80.0    107.0           0

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4238 entries, 0 to 4237
Data columns (total 16 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   male             4238 non-null   int64  
 1   age              4238 non-null   int64  
 2   education        4133 non-null   float64
 3   currentSmoker    4238 non-null   int64  
 4   cigsPerDay       4209 non-null   float64
 5   BPMeds           4185 non-null   float64
 6   prevalentStroke  4238 non-null   int64  
 7   prevalentHyp     4238 non-null   int64  
 8   diabetes         4238 non-null   int64  
 9   totChol          4188 non-null   float64
 10  sysBP            4238 non-null   float64
 11  diaBP            4238 non-null   float64
 12  BMI              4219 non-null   float64
 13  heartRate        4237 non-null   float64
 14  glucose          3850 non-null   float64
 15  TenYearCHD       4238 non-null   int64  
dtypes: float64(9), int64(7)
memory usage: 529.9 KB


How do you load a CSV file into a Pandas DataFrame?
To load a CSV file into a DataFrame, you use the pd.read_csv('file_path') function, where file_path is the path to the CSV file.

What information does the info() function provide about the dataset?
The number of entries (rows) and columns.
The column names.
The number of non-null values in each column.
The data type of each column.
The memory usage of the DataFrame.

How can you identify missing values in the dataset?
To identify missing values, you can use the isnull().sum() function, which gives the count of missing values for each column.


<h3>Exercise 2: Handling Missing Data</h3>


In [12]:
df_dropna = df.dropna()

In [19]:
df_fillna_mean = df.fillna(df.mean())

In [13]:
df_fillna_median = df.fillna(df.median())

In [16]:
df_ffill = df.fillna(method = 'ffill')

/var/folders/yp/cvg_c5fs0gl0gm06r8w_l3p80000gn/T/ipykernel_15611/2525574637.py:1: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df_ffill = df.fillna(method = 'ffill')


In [17]:
df_bfill = df.fillna(method='bfill')

/var/folders/yp/cvg_c5fs0gl0gm06r8w_l3p80000gn/T/ipykernel_15611/500893343.py:1: FutureWarning: DataFrame.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.
  df_bfill = df.fillna(method='bfill')


In [20]:
original_shape = df.shape
dropna_shape = df_dropna.shape
fillna_mean_shape = df_fillna_mean.shape
fillna_median_shape = df_fillna_median.shape
ffill_shape = df_ffill.shape
bfill_shape = df_bfill.shape

In [21]:
original_shape, dropna_shape, fillna_mean_shape, fillna_median_shape, ffill_shape, bfill_shape

((4238, 16), (3656, 16), (4238, 16), (4238, 16), (4238, 16), (4238, 16))

What strategy did you use to handle missing values, and why?
I've used the strategy of dropping nulls using dropn() method

How did filling missing values affect the dataset?
Filling the missing values with the mean, median, or by using forward/backward filling maintained the original size of the dataset (4238 rows), unlike dropping rows, which reduced the dataset to 3656 rows.
Using fillna() methods ensured no data loss but introduced estimated values for the missing data points.

When might it be more appropriate to drop rows with missing values instead of filling them?
Dropping rows is more appropriate when:
The proportion of missing data is small relative to the dataset size.
The missing values occur randomly and don’t follow any pattern.

<h3>Exercise 3: Data Transformation</h3>


In [22]:
from sklearn.preprocessing import MinMaxScaler, StandardScaler

scaler_minmax = MinMaxScaler()
df_minmax_scaled = pd.DataFrame(scaler_minmax.fit_transform(df.select_dtypes(include=['float64', 'int64'])), columns=df.select_dtypes(include=['float64', 'int64']).columns)

scaler_standard = StandardScaler()
df_standard_scaled = pd.DataFrame(scaler_standard.fit_transform(df.select_dtypes(include=['float64', 'int64'])), columns=df.select_dtypes(include=['float64', 'int64']).columns)

df_onehot_encoded = pd.get_dummies(df, columns=['male', 'currentSmoker', 'prevalentStroke', 'prevalentHyp', 'diabetes', 'TenYearCHD'], drop_first=True)

df['age_binned'] = pd.cut(df['age'], bins=[20, 40, 60, 80], labels=['young', 'middle-aged', 'senior'])

df_minmax_scaled_sample = df_minmax_scaled.head()
df_standard_scaled_sample = df_standard_scaled.head()
df_onehot_encoded_sample = df_onehot_encoded.head()
df_age_binned_sample = df[['age', 'age_binned']].head()

df_minmax_scaled_sample, df_standard_scaled_sample, df_onehot_encoded_sample, df_age_binned_sample

(   male       age  education  currentSmoker  cigsPerDay  BPMeds  \
 0   1.0  0.184211   1.000000            0.0    0.000000     0.0   
 1   0.0  0.368421   0.333333            0.0    0.000000     0.0   
 2   1.0  0.421053   0.000000            1.0    0.285714     0.0   
 3   0.0  0.763158   0.666667            1.0    0.428571     0.0   
 4   0.0  0.368421   0.666667            1.0    0.328571     0.0   
 
    prevalentStroke  prevalentHyp  diabetes   totChol     sysBP     diaBP  \
 0              0.0           0.0       0.0  0.149406  0.106383  0.232804   
 1              0.0           0.0       0.0  0.242784  0.177305  0.349206   
 2              0.0           0.0       0.0  0.234295  0.208038  0.338624   
 3              0.0           1.0       0.0  0.200340  0.314421  0.497354   
 4              0.0           0.0       0.0  0.302207  0.219858  0.380952   
 
         BMI  heartRate   glucose  TenYearCHD  
 0  0.277024   0.363636  0.104520         0.0  
 1  0.319680   0.515152  0.101

What is the difference between normalization and standardization?

Normalization (Min-Max Scaling) transforms data to a fixed range, usually between 0 and 1, by subtracting the minimum value and dividing by the range (max-min). This method is sensitive to outliers.
Standardization (Z-score Scaling) transforms data such that it has a mean of 0 and a standard deviation of 1. It is less sensitive to outliers and is useful for algorithms that assume normally distributed data.


How does one-hot encoding transform categorical variables?
One-hot encoding creates a new binary column for each category in the original column. The binary value is 1 if the original observation belongs to that category and 0 otherwise. This method is useful for machine learning algorithms that cannot handle categorical data directly.

Why might you want to bin continuous variables into categories?
Binning continuous variables can simplify data analysis by converting a range of continuous values into discrete categories. This can help with interpretation, especially when looking for trends or segments in the data. Binning is useful for handling skewed data and reducing the impact of outliers.

<h3>Exercise 4: Feature Engineering</h3>


In [23]:
df['BMI_age_interaction'] = df['BMI'] * df['age']

df['age_squared'] = df['age'] ** 2

df_new_features_sample = df[['BMI', 'age', 'BMI_age_interaction', 'age_squared']].head()

df_new_features_sample

BMI  age  BMI_age_interaction  age_squared
0  26.97   39              1051.83         1521
1  28.73   46              1321.58         2116
2  25.34   48              1216.32         2304
3  28.58   61              1743.38         3721
4  23.10   46              1062.60         2116

What new features did you create, and why?
BMI_age_interaction: Created by multiplying BMI with age. This interaction feature might capture the relationship between age and body mass index, which could influence health outcomes in the dataset.
age_squared: Created by squaring the age. This polynomial feature captures any potential non-linear relationship between age and the target variable, especially useful when the effect of age increases disproportionately over time.

How did the new features improve the dataset?
These new features add more complexity to the model by representing interactions or non-linear relationships. In some cases, such transformations can help the model better understand patterns that might not be evident when using the original features alone.

How can date-based features be useful in a dataset?
Date-based features (e.g., year, month, day) can be valuable for understanding seasonal trends, growth patterns, or recurring events. For example:
Year: Identifies long-term trends or changes over time.
Month/Day: Useful for capturing seasonal effects or events that happen periodically within a year.​

<h3>Exercise 5: Data Cleaning</h3>


In [24]:
from scipy import stats

duplicates_before = df.duplicated().sum()
df_cleaned = df.drop_duplicates()
duplicates_after = df_cleaned.duplicated().sum()

z_scores = stats.zscore(df_cleaned.select_dtypes(include=['float64', 'int64']))
abs_z_scores = abs(z_scores)
df_no_outliers = df_cleaned[(abs_z_scores < 3).all(axis=1)]

df_cleaned['male'] = df_cleaned['male'].replace({0: 'female', 1: 'male'})

duplicates_before, duplicates_after, df_no_outliers.shape, df_cleaned[['male']].head()

(0,
 0,
 (0, 19),
      male
 0    male
 1  female
 2    male
 3  female
 4  female)

How did you identify and handle duplicate rows in the dataset?
I used the duplicated() function to check for duplicate rows. In this dataset, there were no duplicates found (0 before and after applying drop_duplicates()), so no rows were removed.

What method did you use to detect and remove outliers, and why?
The Z-score method was used to detect outliers. This method calculates the Z-score for each value, which measures how many standard deviations a value is from the mean. A common threshold of 3 was applied, meaning any data points with a Z-score greater than 3 or less than -3 were considered outliers. In this dataset, all rows were considered outliers and removed due to extreme values.

How did you address inconsistencies in categorical data?
Although the dataset did not contain textual categorical inconsistencies, the male column (initially represented as 0 or 1) was standardized by replacing 0 with 'female' and 1 with 'male'. This is an example of how to handle categorical inconsistencies by ensuring consistent encoding across the dataset.

<h3>Exercise 6: Splitting Data into Training and Testing Sets</h3>


In [25]:
from sklearn.model_selection import train_test_split

X = df_cleaned.drop('TenYearCHD', axis=1)
y = df_cleaned['TenYearCHD']

X_train_70, X_test_70, y_train_70, y_test_70 = train_test_split(X, y, test_size=0.3, random_state=42)

X_train_80, X_test_80, y_train_80, y_test_80 = train_test_split(X, y, test_size=0.2, random_state=42)

train_test_split_shapes = {
    "70-30 Train Set": X_train_70.shape,
    "70-30 Test Set": X_test_70.shape,
    "80-20 Train Set": X_train_80.shape,
    "80-20 Test Set": X_test_80.shape
}

train_test_split_shapes

{'70-30 Train Set': (2966, 18),
 '70-30 Test Set': (1272, 18),
 '80-20 Train Set': (3390, 18),
 '80-20 Test Set': (848, 18)}

How do you split a dataset into training and testing sets in Python?
You can use the train_test_split() function from sklearn.model_selection. This function randomly splits the dataset into training and testing sets, specifying the test size as a proportion of the total data.


What considerations should you keep in mind when choosing a train-test split ratio?
Data size: With a small dataset, a larger test set may leave too little data for training, causing the model to underfit.
Complexity of the model: More complex models may require a larger training set to capture patterns in the data.
Desired generalization: The split should leave enough data in the test set to properly evaluate the model's generalization ability.

How does the size of the training set impact the model's ability to generalize?
A larger training set generally helps the model learn better patterns from the data, reducing the chance of overfitting. However, if the training set is too large, the test set may be too small, resulting in unreliable performance evaluation. Conversely, if the training set is too small, the model might not capture enough information to generalize well. ​

<h3>Exercise 7: Data Preprocessing Pipeline</h3>


In [27]:
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.ensemble import RandomForestClassifier

numerical_features = ['age', 'cigsPerDay', 'totChol', 'sysBP', 'diaBP', 'BMI', 'heartRate', 'glucose']
categorical_features = ['male', 'currentSmoker', 'prevalentStroke', 'prevalentHyp', 'diabetes']

numerical_pipeline = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='mean')), 
    ('scaler', StandardScaler())  
])

categorical_pipeline = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),  
    ('onehot', OneHotEncoder(drop='first'))  
])

preprocessor = ColumnTransformer(
    transformers=[
        ('num', numerical_pipeline, numerical_features),
        ('cat', categorical_pipeline, categorical_features)
    ])

model_pipeline = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('classifier', RandomForestClassifier(random_state=42))
])

model_pipeline.fit(X_train_70, y_train_70)

y_pred_70 = model_pipeline.predict(X_test_70)

transformed_test_data_sample = X_test_70.head()
predictions_sample = y_pred_70[:5]

transformed_test_data_sample, predictions_sample

(        male  age  education  currentSmoker  cigsPerDay  BPMeds  \
 3188    male   63        1.0              0         0.0     0.0   
 764     male   45        3.0              0         0.0     0.0   
 3264  female   51        1.0              1         2.0     0.0   
 1967    male   45        3.0              1        30.0     0.0   
 2185  female   45        2.0              1         3.0     0.0   
 
       prevalentStroke  prevalentHyp  diabetes  totChol  sysBP  diaBP    BMI  \
 3188                0             1         0    190.0  148.0   90.0  27.13   
 764                 0             0         0    162.0  125.0   89.0  27.98   
 3264                0             0         0    261.0  127.0   81.0  20.24   
 1967                0             0         0    250.0  126.0   89.5  28.68   
 2185                0             0         0    250.0  130.0   80.0  20.24   
 
       heartRate  glucose   age_binned  BMI_age_interaction  age_squared  
 3188       72.0     86.0       s

What are the benefits of using a preprocessing pipeline?
Automation: A pipeline automates the preprocessing steps, ensuring they are applied consistently and in the correct order.
Modularity: Each step (e.g., imputation, scaling, encoding) is handled separately, making the process more manageable and easier to modify or extend.

How does the pipeline ensure consistency between training and test data transformations?
When a pipeline is fit to the training data, it learns the parameters (e.g., mean for imputation, scaling factors) from the training set. These same parameters are then applied to the test data, ensuring consistent transformations across both sets.

How can you extend the pipeline to include additional preprocessing steps?
Additional steps, such as feature engineering, outlier removal, or custom transformations, can be added by extending the pipeline with additional Pipeline objects or custom transformers. This can be done by simply adding new steps to the pipeline using the appropriate transformation functions or classes.